src/globus_sdk/experimental/globus_app/_validating_storage_adapater.py

```diff
@@ -0,0 +1,213 @@
+from __future__ import annotations
+
+import time
+import typing as t
+
+from globus_sdk import AuthClient, OAuthTokenResponse, Scope
+from globus_sdk.experimental.consents import ConsentForest
+from globus_sdk.tokenstorage import StorageAdapter
+
+from ..._types import UUIDLike
+from ._identifiable_oauth_token_response import (
+    IdentifiedOAuthTokenResponse,
+    expand_id_token,
+)
+from .errors import (
+    ExpiredTokenError,
+    IdentityMismatchError,
+    UnmetScopeRequirementsError,
+)
+
+
+class ValidatingStorageAdapter(StorageAdapter):
+    """
+    A special version of a StorageAdapter which wraps another storage adapter and
+        validates that tokens meet certain requirements when storing/retrieving them.
+
+    The adapter is not concerned with the actual storage location of tokens but rather
+        validating that they meet certain requirements:
+        1) Identity Requirements
+            a) Identity info is present in the token data (this requires that the
+                token data was retrieved with the "openid" scope in addition to any
+                other scope requirements).
+            b) The identity info in the token data matches the identity info stored
+                previously in the adapter.
+        2) Scope Requirements
+            b) Each newly polled resource server's token meets the root scope
+                requirements for that resource server.
+            c) Polled consents meets all dependent scope requirements.
+    """
+
+    def __init__(
+        self,
+        storage_adapter: StorageAdapter,
+        scope_requirements: dict[str, list[Scope]],
+        *,
+        consent_client: AuthClient | None = None,
+    ):
+        """
+        :param storage_adapter: The storage adapter being wrapped.
+        :param scope_requirements: A collection of resource-server keyed scope
+            requirements to validate on token storage/retrieval.
+        :param consent_client: An AuthClient to be used for consent polling. If omitted,
+            dependent scope requirements are ignored during validation.
+        """
+        self._storage_adapter = storage_adapter
+        self.scope_requirements = scope_requirements
+        self._consent_client = consent_client
+
+        self.identity_id = self._lookup_stored_identity_id()
+        self._cached_consent_forest = self._poll_and_cache_consents()
```

Comment:
Would it make sense to store the cached consent forest similarly to the identity id? Otherwise I think the CLI would be making a `get_consents` call for every command that has dependent scope requirements

Reply:
Talked offline; yes but not in this unit of work.